ЛР 2 Вар 4 (Порушенков) ВТ-119

Vitaliy308 opened the pull request:
Не реализован кольцевой буфер и первая часть индивидуального задания (проверка вхождения одной структуры в другую посредством циклического сдвига элементов)

GOOD-Stuff commented on lab_03/src/binary_tree.h:
```diff
@@ -0,0 +1,323 @@
+        Element* elem = root;
+        while (elem != nullptr){
+            if (d < elem->data)
+                elem = current->Left;
```
Не компилируется. 
`current` не объявлен в методе.

GOOD-Stuff commented on lab_03/src/binary_tree.h:
```diff
@@ -0,0 +1,323 @@
+        ~Element() = default;
+    };
+    Element* root { nullptr };
+    vector<T> z;
+    vector<int> p;
```
Что за `z` и `p`?
За что отвечают эти вектора? Именуйте сущности яснее.

GOOD-Stuff commented on lab_03/src/binary_tree.h:
```diff
@@ -0,0 +1,323 @@
+
+    int heightElement(Element* elm){
+        if (elm == nullptr) return 0;
+        return max(heightElement(elm->Right),heightElement(elm->Left))+1;
```
Почему, если для одного узла (корня) вы возвращаете высоту равную 0, то для N узлов, уже вернёте max(height(N)) + 1? 
+1 лишний.

GOOD-Stuff commented on lab_03/src/general_tree.h:
```diff
@@ -0,0 +1,280 @@
+    vector<Element*> Elem;
+    vector<int> visota;
+
+    void DeleteElements (Element* elm) {
```
Кажется вы тут забыли освободить память из под узлов.

GOOD-Stuff commented on lab_03/src/general_tree.h:
```diff
@@ -0,0 +1,280 @@
+    void DeleteElements (Element* elm) {
+        if (elm != nullptr ){
+            for (int i = 0; i < elm->Subtrees.size(); i++)
+                DeleteElements(elm->Subtrees.begin()[i]);
```
Какой-то странный у вас способ обращения к элементу.
Используйте в таком случае либо полноценные итераторы, либо метод `at(i)`, либо `[i]`.

Vitaliy308 committed:
completing

Не реализован кольцевой буфер и первая часть индивидуального задания (проверка вхождения одной структуры в другую посредством циклического сдвига элементов)

GOOD-Stuff commented:
**Всегда** в названии своих реквестов **указывайте**: 
- группу
- номер работы
- вариант
- фамилии участников

иначе как мне понять, кто вы и что вы сдаёте?!

GOOD-Stuff commented:
> Не реализован кольцевой буфер и первая часть индивидуального задания

а они вообще будут реализовываться?

Vitaliy308 commented:
- ВТ-119
- ЛР 2
- Вариант 4
- Порушенков Виталий

Нет, скорее всего не будут реализовываться

GOOD-Stuff commented on src/DoubleLinkedList.h:
```diff
@@ -0,0 +1,374 @@
+		if (!a) return b;
+		if (!b) return a;
+
+		Item* c = 0;
```
указатели приводите к `nullptr`, вместо 0 и `NULL`.

GOOD-Stuff commented on src/DoubleLinkedList.h:
```diff
@@ -0,0 +1,374 @@
+	}
+
+	void pop_front(){
+		Item* oldStart = pStart;
```
Отсутствует проверка на попытку выталкивания из пустой структуры

GOOD-Stuff commented on src/DoubleLinkedList.h:
```diff
@@ -0,0 +1,374 @@
+		}
+
+	public:
+		iterator() {}
```
такие конструкторы объявляйте как `default`: `iterator() = default`.

GOOD-Stuff commented on src/DoubleLinkedList.h:
```diff
@@ -0,0 +1,374 @@
+		return (*this > list) || (*this == list);
+	}
+
+	void merge_sort(const DoubleLinkedList<T>& list){
```
А это что/зачем?

GOOD-Stuff commented on src/DoubleLinkedList.h:
```diff
@@ -0,0 +1,374 @@
+	};
+
+	iterator begin() const{
+		return iterator(pStart, const_cast<DoubleLinkedList<T>*>(this));
```
Не используйте `const_cast`

GOOD-Stuff reviewed (commented):
За работу ставлю 7
